// @react-md/states/lib/ripples/Ripple.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ripple = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var classnames_1 = __importDefault(require("classnames"));
var transition_1 = require("@react-md/transition");
var StatesConfig_1 = require("../StatesConfig");
function Ripple(_a) {
    var className = _a.className, propClassNames = _a.classNames, propTimeout = _a.timeout, ripple = _a.ripple, entered = _a.entered, exited = _a.exited;
    var exiting = ripple.exiting, style = ripple.style;
    var timeout = propTimeout;
    var classNames = propClassNames;
    var context = (0, StatesConfig_1.useStatesConfigContext)();
    if (typeof timeout === "undefined" || typeof classNames === "undefined") {
        if (typeof timeout === "undefined") {
            timeout = context.rippleTimeout;
        }
        if (typeof classNames === "undefined") {
            classNames = context.rippleClassNames;
        }
    }
    var _b = (0, transition_1.useCSSTransition)({
        appear: true,
        transitionIn: !exiting,
        timeout: timeout,
        className: (0, classnames_1.default)("rmd-ripple", className),
        classNames: classNames,
        onEntered: function () {
            entered(ripple);
        },
        onExited: function () {
            exited(ripple);
        },
    }), elementProps = _b.elementProps, rendered = _b.rendered;
    if (!rendered) {
        return null;
    }
    return (0, jsx_runtime_1.jsx)("span", __assign({}, elementProps, { style: style }));
}
exports.Ripple = Ripple;
