// @react-md/table/lib/TableRow.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableRow = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("react");
var classnames_1 = __importDefault(require("classnames"));
var utils_1 = require("@react-md/utils");
var config_1 = require("./config");
var block = (0, utils_1.bem)("rmd-tr");
/**
 * Creates a `<tr>` element with some general styles that are inherited from the
 * base table configuration.
 */
exports.TableRow = (0, react_1.forwardRef)(function TableRow(_a, ref) {
    var className = _a.className, propDisableHover = _a.disableHover, propDisableBorders = _a.disableBorders, children = _a.children, _b = _a.selected, selected = _b === void 0 ? false : _b, _c = _a.clickable, clickable = _c === void 0 ? false : _c, props = __rest(_a, ["className", "disableHover", "disableBorders", "children", "selected", "clickable"]);
    var _d = (0, config_1.useTableConfig)({
        disableHover: propDisableHover,
        disableBorders: propDisableBorders,
    }), disableHover = _d.disableHover, disableBorders = _d.disableBorders;
    return ((0, jsx_runtime_1.jsx)("tr", __assign({}, props, { ref: ref, className: (0, classnames_1.default)(block({
            bordered: !disableBorders,
            hoverable: !disableHover,
            clickable: clickable,
            selected: selected,
            "selected-hoverable": selected && !disableHover,
        }), className) }, { children: children })));
});
