// @react-md/expansion-panel/lib/ExpansionPanel.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpansionPanel = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("react");
var classnames_1 = __importDefault(require("classnames"));
var card_1 = require("@react-md/card");
var transition_1 = require("@react-md/transition");
var utils_1 = require("@react-md/utils");
var ExpansionPanelHeader_1 = require("./ExpansionPanelHeader");
var block = (0, utils_1.bem)("rmd-expansion-panel");
/**
 * The expansion panel renders a header element (that is just a button) and
 * dynamically shows content once expanded.
 */
exports.ExpansionPanel = (0, react_1.forwardRef)(function ExpansionPanel(_a, ref) {
    var className = _a.className, children = _a.children, headerStyle = _a.headerStyle, headerClassName = _a.headerClassName, contentStyle = _a.contentStyle, contentClassName = _a.contentClassName, headerRef = _a.headerRef, _b = _a.disablePadding, disablePadding = _b === void 0 ? false : _b, _c = _a.disableSecondaryColor, disableSecondaryColor = _c === void 0 ? false : _c, customHeader = _a.customHeader, header = _a.header, expanded = _a.expanded, onExpandClick = _a.onExpandClick, expanderIcon = _a.expanderIcon, _d = _a.marginTop, marginTop = _d === void 0 ? false : _d, _e = _a.fullWidth, fullWidth = _e === void 0 ? true : _e, _f = _a.persistent, persistent = _f === void 0 ? false : _f, _g = _a.disabled, disabled = _g === void 0 ? false : _g, _h = _a.disableTransition, disableTransition = _h === void 0 ? false : _h, props = __rest(_a, ["className", "children", "headerStyle", "headerClassName", "contentStyle", "contentClassName", "headerRef", "disablePadding", "disableSecondaryColor", "customHeader", "header", "expanded", "onExpandClick", "expanderIcon", "marginTop", "fullWidth", "persistent", "disabled", "disableTransition"]);
    var id = props.id;
    var contentId = "".concat(id, "-content");
    var _j = (0, transition_1.useCollapseTransition)({
        style: contentStyle,
        className: contentClassName,
        timeout: disableTransition ? 0 : undefined,
        transitionIn: expanded,
        temporary: !persistent,
    }), elementProps = _j.elementProps, rendered = _j.rendered;
    return ((0, jsx_runtime_1.jsxs)(card_1.Card, __assign({}, props, { id: "".concat(id, "-container"), ref: ref, fullWidth: fullWidth, className: (0, classnames_1.default)(block({ expanded: expanded, "margin-top": marginTop }), className) }, { children: [customHeader || ((0, jsx_runtime_1.jsx)(ExpansionPanelHeader_1.ExpansionPanelHeader, __assign({ "aria-disabled": disabled || undefined, id: id, ref: headerRef, style: headerStyle, className: headerClassName, icon: expanderIcon, expanded: expanded, onClick: onExpandClick, disableTransition: disableTransition }, { children: header }))), rendered && ((0, jsx_runtime_1.jsx)(card_1.CardContent, __assign({}, elementProps, { id: contentId, "aria-labelledby": id, role: "region", disableSecondaryColor: disableSecondaryColor, disablePadding: disablePadding, disableExtraPadding: true }, { children: children })))] })));
});
